Querier: propagate errors with status code 422 coming from the store-gateway

duricanikolic opened the pull request:
#### What this PR does
This PR introduces the following changes:
- Makes the querier propagate errors with status code 422 coming from the store-gateway
- New integration tests have been added in store_gateway_limits_hit_test.go: they ensure that when the max-fetched-chunks-per-query or max-fetched-series-per-query limits are hit, both store-gateway and querier will return a 422 status code
#### Which issue(s) this PR fixes or relates to

Extension of the fix for issue #3382 

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented:
> store_gateway_client.go and store_gateway_test.go have been moved from the querier to the storegatewaypb package

The `*pb` packages should contain only protobufs.

That being said I don't think an integration should use a gRPC client. The idea of integration tests is that they should emulate how users use the system. Let's talk on Slack about it, so we can see if we can find a better approach for the testing here.

duricanikolic commented:
> The *pb packages should contain only protobufs.
@pracucci I have fixed that and moved the 2 sources to the storegateway package instead. 

For the rest, let's discuss it next week after you have reviewed the code.

56quarters commented on pkg/storegateway/store_gateway_client.go:
```diff
@@ -3,7 +3,7 @@
 // Provenance-includes-copyright: The Cortex Authors.
 
-package querier
+package storegateway
```
WDYT about putting this in `storegateway/client`, same as we do for the ingester client?

duricanikolic replied:
@56quarters I had a call with @colega in which we discussed that probably the whole implementation of [store_gateway_client.go](https://github.com/grafana/mimir/blob/main/pkg/querier/store_gateway_client.go) (currently belonging to the `querier`, and not `storegateway` package), as well as of [pool.go](https://github.com/grafana/dskit/blob/main/ring/client/pool.go) were done before we started using generics, and therefore the implementation looked a bit more complicated than it should be. The changes I made in `store_gateway_client.go` were needed for a new test that I introduced, and that will now be removed. So, I will also remove the changes I did there. Nevertheless, we find that it might be appropriate to introduce generics in `pool.go` and related sources, and to update mimir accordingly. What do you think about that?

56quarters commented on pkg/storegateway/store_gateway_client.go:
```diff
@@ -47,32 +47,32 @@ func dialStoreGatewayClient(clientCfg grpcclient.Config, addr string, requestDur
 }
 
-type storeGatewayClient struct {
+type ClientImpl struct {
```
If this is implementing an interface, we could refer to the interface everywhere and make the actual struct unexported.

duricanikolic replied:
Will be fixed in a separate PR

56quarters commented on pkg/storegateway/store_gateway_client.go:
```diff
@@ -47,32 +47,32 @@ func dialStoreGatewayClient(clientCfg grpcclient.Config, addr string, requestDur
 }
 
-func (c *storeGatewayClient) Close() error {
+func (c *ClientImpl) Close() error {
```
`*Impl` feels like kind of a Java-ism. Doesn't seem like we use it anywhere else in the Mimir codebase.

56quarters commented on pkg/querier/blocks_store_replicated_set_test.go:
```diff
@@ -24,6 +24,8 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
+	"github.com/grafana/mimir/pkg/storegateway"
```
All the mimir imports should be grouped together

duricanikolic replied:
Will be fixed in a separate PR

56quarters commented on integration/store_gateway_limits_hit_test.go:
```diff
@@ -0,0 +1,202 @@
+	rangeResultResponse, rangeResultBody, err := client.QueryRangeRaw("{__name__=~\"series_.+\"}", timeStamps12[series1], timeStamps34[series4].Add(1*time.Hour), time.Second)
+	require.NoError(t, err)
+	require.Equal(t, http.StatusUnprocessableEntity, rangeResultResponse.StatusCode)
+	require.JSONEq(t, `{"status":"error","errorType":"execution","error":"expanding series: the query exceeded the maximum number of series (limit: 3 series) (err-mimir-max-series-per-query). To adjust the related per-tenant limit, configure -querier.max-fetched-series-per-query, or contact your service administrator."}`, string(rangeResultBody))
```
Instead of requiring that the JSON match exactly, seems sufficient to just make sure it contains the expected global error name (`err-mimir-max-series-per-query`).

pracucci commented on integration/store_gateway_limits_hit_test.go:
```diff
@@ -0,0 +1,202 @@
+			{Type: storepb.LabelMatcher_RE, Name: labels.MetricName, Value: "series_.+"},
+		},
+	}
+	stream, err := storeGatewayClient.(*storegateway.ClientImpl).Series(ctx, req)
+	require.NoError(t, err)
+	require.NotNil(t, stream)
```
Integration tests should focus on end-to-end behaviour. We shouldn't call store-gateway via gRPC, but run a normal query and assert on the actual result / error we get.

In this test we want to focus on the limit hit in the store-gateway. I think the solution is easy: just set the limit `-querier.max-fetched-series-per-query` to the store-gateway and **not** the querier. Since it's passed as a CLI flag, you can pass it to the store-gateway but not the querier.

In the test testing this limit in the querier, you can pass it to the querier but not the store-gateway.

The two tests ("hit limit in the querier" and "hit limit in the store-gateway") can also be written in the same `TestXXX()` function, because they should be the exact same, but passing the limit to one component or the other (so a normal table test should be fine).

pracucci commented on integration/store_gateway_limits_hit_test.go:
```diff
@@ -0,0 +1,225 @@
+
+const (
+	blockRangePeriod = 5 * time.Second
+	ingesterTag      = "ingester"
+	distributorTag   = "distributor"
+	storeGatewayTag  = "store-gateway"
+	querierTag       = "querier"
+	orgID            = "test"
+	series1          = "series_1"
+	series2          = "series_2"
+	series3          = "series_3"
+	series4          = "series_4"
```
[nit] I would remove all these constants. Don't think give much value, if you accept the comment I left on `createClient()`.

duricanikolic replied:
I did accept the comment about `createClient()` and removed some of the constants. I have left a couple of them that are used more times, in order to avoid hardcoding.

pracucci commented on integration/store_gateway_limits_hit_test.go:
```diff
@@ -0,0 +1,225 @@
+	ingester := e2emimir.NewIngester(ingesterTag, consul.NetworkHTTPEndpoint(), flags)
+	mimirServices[ingesterTag] = ingester
+
+	if len(additionalStoreGatewayFlags) == 0 {
```
[nit] To simplify it, I think we can always call `mergeFlags()`, no?

pracucci commented on integration/store_gateway_limits_hit_test.go:
```diff
@@ -0,0 +1,225 @@
+
+	require.NoError(t, scenario.StartAndWaitReady(distributor, ingester, storeGateway))
+
+	if len(additionalQuerierFlags) == 0 {
```
[nit] Same here.

pracucci commented on integration/store_gateway_limits_hit_test.go:
```diff
@@ -0,0 +1,225 @@
+	}
+}
+
+func Test_MaxChunksPerQueryLimitHit(t *testing.T) {
```
[nit] I think you can merge the two tests together. Other comments I left assume we do merge the two tests.

pracucci reviewed (approved):
Good job! The fix LGTM. I left few comments about the integration test cause I think we can further simplify it.

pracucci commented on integration/store_gateway_limits_hit_test.go:
```diff
@@ -0,0 +1,160 @@
+		require.NoError(t, scenario.StartAndWaitReady(distributor, ingester, storeGateway))
+
+		additionalQuerierFlags = mergeFlags(flags, additionalQuerierFlags)
+		querier := e2emimir.NewQuerier(querierTag, consul.NetworkHTTPEndpoint(), additionalQuerierFlags)
```
[nit] The `querierTag` is only used here. Can be removed.

duricanikolic replied:
I missed this one. Sorry for this.

pracucci commented on integration/store_gateway_limits_hit_test.go:
```diff
@@ -0,0 +1,160 @@
+
+		require.NoError(t, scenario.StartAndWaitReady(querier))
+
+		client, err = e2emimir.NewClient(distributor.HTTPEndpoint(), querier.HTTPEndpoint(), "", "", orgID)
```
[nit] `orgID` constant is only used here. Can be removed.

pracucci commented on integration/store_gateway_limits_hit_test.go:
```diff
@@ -0,0 +1,160 @@
+	blockRangePeriod = 5 * time.Second
+	querierTag       = "querier"
+	orgID            = "test"
+	series1          = "series_1"
+	series2          = "series_2"
+	series3          = "series_3"
+	series4          = "series_4"
```
[nit] The series constants are not required be defined globally. Can be moved inside `Test_MaxSeriesAndChunksPerQueryLimitHit`.

pracucci commented on integration/store_gateway_limits_hit_test.go:
```diff
@@ -0,0 +1,160 @@
+			scenario, ingester, storeGateway, client := setup(t, testData.additionalStoreGatewayFlags, testData.additionalQuerierFlags)
+			defer scenario.Close()
+
+			timeStamps12, timeSeries12 := createTimeSeries(t, []string{series1, series2})
```
[nit] Can we create all timeseries upfront here, instead of calling `createTimeSeries()` twice (here and below)? If yes, then what's the need of `createTimeSeries()` at all?

pracucci reviewed (commented):
Thanks for your patience. Left few more comments.

pracucci commented on integration/store_gateway_limits_hit_test.go:
```diff
@@ -0,0 +1,145 @@
+)
+
+const (
+	blockRangePeriod = 5 * time.Second
```
[nit] Final nit, as promised. `blockRangePeriod` is only used in the test funciton. You can move the constant there. Keep in mind you can define `const` inside the function itself. Just no need for a global const, which may be misused from other tests.

duricanikolic replied:
Hi @pracucci, 
Could you please do another check and then merge this PR if it looks OK to you?
Thank you

pstibrany commented on CHANGELOG.md:
```diff
@@ -4,7 +4,7 @@
 ### Grafana Mimir
 
-* [CHANGE] Querier: returning 422 when query hits `max_fetched_chunks_per_query` and `max_fetched_series_per_query` limits in the store-gateway. #4056
+* [CHANGE] Querier: returning 422 when query hits `max_fetched_chunks_per_query` and `max_fetched_series_per_query` limits either in the store-gateway or in querier. #4056 #4100
```
Can you please explain what has changed for this [CHANGE]?

duricanikolic replied:
Hi @pstibrany, 
Until this change querier was, probably erroneously, reporting a status check error in the case the limit was hit in the store-gateway. There are 2 PRs for the following reason:
- PR 4056: when a limit is hit in Limiter, an error with the status code 422 is created and returned. These errors are the ones that the store-gateway returns to the querier
- PR 4100: the status code 422 returned by the store-gateway is further propagated by the querier, and not converted to a consistency check anymore

pstibrany replied:
Would you please update CHANGELOG file to explain that? Right now the changelog entry says `[CHANGE]` but doesn't say what has changed, it only mentions what the code does now. As a user reading this entry, I don't understand why it's under `[CHANGE]`.

duricanikolic replied:
Sure, I have updated CHANGELOG
